Utilizing tiddlywiki caching

flibbles opened the pull request:
By utilizing tiddlywiki's inbuilt caching mechanism, it spares the ViewAbstraction from having to manage one itself. Hopefully, but letting Tiddlywiki manage each individual view tiddler on its own, it can reduce the amount of refreshing.

I'm currently using TiddlyMap on my machine with these changes active. I haven't encountered any problems, but I'm still only submitting this PR at this time with intent for review.

This is my proposal for #286

Comments are among the changes.

flibbles commented on src/plugins/felixhayashi/tiddlymap/js/widget/MapWidget.js:
```diff
@@ -1931,8 +1931,8 @@ class MapWidget extends Widget {
+    const nodeData = Object.assign({}, this.view.getNodeData(node.id));
     // we need to delete the positions so they are not reset when a user
     // resets the style…
     delete nodeData.x;
```
This is one change I'm not completely sure about. I'm not sure what it meant with "need a new object". It looks like the MapWidget wants to modify the nodeData, and that's why?

I've switched it to shallow cloning the nodeData so it can remove `nodeData.x` and `nodeData.y` without affecting the underlying cached object. Because there shouldn't be any case where the cached object itself is modified.

In fact, I was considering deep freezing all of ViewAbstraction's cached objects. Tiddlywiki does the same with tiddlers. They're never modified except by addTiddler, which refreshes the cache.

felixhayashi replied:
> I'm not sure what it meant with "need a new object". It looks like the MapWidget wants to modify the nodeData, and that's why?
yeah, I want to persist the style but I delete the x,y coordinates before persisting. If I'd do that operation on the actual nodeData that is used in the view, the coordinates would be modified and lost. so I created a fresh copy via the `true` param in `getNodeData(node.id, true)` and do the operations on this object.

Your solution is fine for the x and y case, because the int primitive types are copied by value, not by reference. yet, if we'd ever needed to delete properties of nested objects in `nodeData`, the Object.assign would not suffice anymore and this would cause side effects by deleting nested props. Anyhow, I am fine with this change, please add a todo though `// note: a deep copy would be needed if nested properties were modified`.

also please use this syntax: `{ ...this.view.getNodeData(node.id) }`. thanks :)

flibbles commented on src/plugins/felixhayashi/tiddlymap/js/graph/ViewAbstraction.js:
```diff
@@ -358,21 +344,20 @@ class ViewAbstraction {
-      this.config
-    ));
-
   }
```
Moved this up into a narrower scope where `config` is defined. If you follow the code, it looks like having the `addTiddler` at the very end just had it get called extraneously during `setConfig({object})` calls.

felixhayashi replied:
I agreee, well spotted.

flibbles commented on src/plugins/felixhayashi/tiddlymap/js/graph/ViewAbstraction.js:
```diff
@@ -81,23 +78,17 @@ class ViewAbstraction {
+    return (updates[env.path.edgeTypes] || utils.hasKeyWithPrefix(changedTiddlers, this.getRoot()));
 
   }
 
```
I considered changing the name of `update` to `updated` or `modified` or something, since it's now a zero-side-effect method. I think there might be some simplification possible in MapWidget now, where this is called.

felixhayashi replied:
I understand your intention. In this case I would rename it to `hasUpdated()`. On the other hand, leaving it as a function with side effects also leaves the chance for us/others to add side effects in the future. This is why I called it `update` in analogy to the widget's `update` method. What would you suggest?

flibbles replied:
If MapWidget is the only place where it's called, I'd recommend changing it. It can always be changed back later, since it's only two lines of code changed.

It _is_ only used in MapWidget, right? I couldn't find it anywhere else.

felixhayashi replied:
> It is only used in MapWidget, right? I couldn't find it anywhere else.

yes

> If MapWidget is the only place where it's called, I'd recommend changing it. It can always be changed back later, since it's only two lines of code changed.

ok!

felixhayashi commented on src/plugins/felixhayashi/tiddlymap/js/graph/ViewAbstraction.js:
```diff
@@ -284,24 +274,18 @@ class ViewAbstraction {
 
-    if (!isRebuild && this.config) {
+    const self = this;
+    const config = $tw.wiki.getCacheForTiddler(this.configTRef, "tmap-config", function() {
```
please use an arrow function and remove `self = this` above. the arrow function automatically binds to `this`

felixhayashi commented on src/plugins/felixhayashi/tiddlymap/js/graph/ViewAbstraction.js:
```diff
@@ -523,35 +502,31 @@ class ViewAbstraction {
-
-    } else {
+    const self = this;
+    const filter = $tw.wiki.getCacheForTiddler(this.edgeTypeFilterTRef, "tmap-edgeTypeFilter", function() {
```
please also an arrow function here instead of using self

felixhayashi commented on src/plugins/felixhayashi/tiddlymap/js/graph/ViewAbstraction.js:
```diff
@@ -610,14 +581,14 @@ class ViewAbstraction {
-      ? this.nodeData
-      : utils.parseFieldData(this.mapTRef, 'text', {}));
+    const self = this;
+    const data = $tw.wiki.getCacheForTiddler(this.mapTRef, "tmap-map", function() {
```
arrow function () => utils.parseFieldData(self.mapTRef, 'text', {});

felixhayashi commented:
Hey @flibbles, sorry for the delay. Please consider my comments, it's only about some minor code style changes. If you say you tested it, we can merge it!

flibbles commented:
I have switched to using lambda functions in all the listed cases.

I've switched to `{ ...this.view.getNodeData(node.id) }` in that one case, which by the way, [_completely_ blew my mind](https://media2.giphy.com/media/xT0xeJpnrWC4XWblEk/giphy.gif). I never even knew that kind of syntax existed in javascript. I thought that kind of stuff was limited to ruby or lisp or other _progressive_ languages.

I left a comment recommending `$tw.utils.deepCopy` should the need ever arise.

Other than that, there's just the issue of renaming `update` to `hasUpdated`. I put the change in, and as long as it's only used in MapWidget, we're good. No problems running this stuff on my project.

felixhayashi commented:
> which by the way, completely blew my mind. I never even knew that kind of syntax existed in javascript. I thought that kind of stuff was limited to ruby or lisp or other progressive languages.

Yeah, these pretty changes came with ES2015 :tada: 

> Other than that, there's just the issue of renaming update to hasUpdated. I put the change in, and as long as it's only used in MapWidget, we're good. No problems running this stuff on my project.

from my side everything looks fine. nice job squashing the commit after the review! will merge it now :)
